### ISSUE TITLE
lag detection: (was controls must respond when clicked/used)

### ISSUE THREAD

**Issue by ctm** [ctm/mb2-doc#160]
Currently when someone clicks on a check-box, the client sends the relevant request to the server and we don't actually update the UI until the server responds. That seemed like a good idea at the time and it probably works > 99% of the time. However, if someone has network connectivity issues, that shows up as unresponsive UI elements which is sufficiently frustrating that the player may not even realize there are network problems due to the apparent failure of the UI.

I need to improve advanced action anyway (#127), including adding some synchronization info on the request so that the server can recognize lagged actions and not act on them. I should take the requirement of this issue into account when I work on that one.

The nice thing is that these two tickets aren't particularly hard to do. The frustrating thing is they're still relatively low priority due to a backlog of work in part due to me not spinning up a front-end person yet.

**Comment by ctm** [comment]
I don't know if I'll actually work on this one soon, but I'm marking this as high priority because it really makes a bad internet connection worse. It is especially confusing to people who either haven't been told what is going on or who have been told repeatedly and either don't understand or don't internalize or don't remember it.

**Comment by ctm** [comment]
AIYAH❗had an important insight. There's a chance that Llew's trackpad sends double-clicks in some situations and if so, that would hit this bug much more often than it gets hit due to network hiccups.

I should see if I can find a way to configure my (Mac) laptop to send double-clicks. Assuming I can do that, I should then see if I can reproduce some of the problems Llew (and occasionally others) has had off and on.

**Comment by ctm** [comment]
I removed the `high priority` label because I've not touched this issue in a long time. My creation of a graphical interface has taken precedence and the resultant restructuring needs to be done before I tackle this one anyway.

**Comment by ctm** [comment]
This continues to bite people and it's poor form that I stripped the `high priority` label and then let this languish. Here's a nicely detailed report from a player:

> I would be waiting for Sam6 to act, and would be watching the timing bar move under his name, and nothing would happen. I'd try clicking various buttons and wouldn't get a response. I would try messaging on the client messenger, and it wouldn't work. I could get the messenger on the home screen to work, as you see.

> Eventually the client would come to life and I would see that I had timed out and had missed several hands.

> What I assume happened was that I dropped internet service, or that it slowed down drastically, while the action was on Sam6, and the client didn't interpret that I was dropped. 

> I have noted previously that I can be timing out with the action on me, and I'll be watching my timing bar move, but unable to get any button pushes to activate. I have assumed that the timing bar acts more as a pacifier screen and less as real-time feedback.

> This doesn't explain why I was able to message on the home screen. It could be a result of a slow or intermittent connection that required less bandwidth.

> At the time this happened, I was playing through a VPN on an airport connection, and although I wasn't having other problems, the whole thing seemed a little squirrelly.

**Comment by ctm** [comment]
This bit tegwin last night, too. From her POV mb2 was putting her on vacation before it was even her turn. From the other players' POV and from the server, she was given 20 seconds to act and didn't, so it put her on vacation:
```
select id, hand_id, received_at, message from public_table_messages where table_id = 5213 and player_id IS NULL order by received_at asc;
...
 9244075 |  354627 | 2023-06-29 00:12:10.305587+00 | {"Remind": [20000, 0, "2023-06-29T00:12:30.285937619Z", 30, 20]}
 9244086 |  354627 | 2023-06-29 00:12:16.850437+00 | {"Bets": [30, 800, null, false]}
 9244087 |  354627 | 2023-06-29 00:12:16.882639+00 | {"Remind": [20000, 800, "2023-06-29T00:12:36.849088206Z", 19, 20]}
 9244093 |  354627 | 2023-06-29 00:12:36.851904+00 | {"Vacationed": 19}
 9244094 |  354627 | 2023-06-29 00:12:36.85644+00  | {"Folds": [19, true, null, "Vacation"]}
 9244095 |  354627 | 2023-06-29 00:12:36.860925+00 | {"Remind": [20000, 800, "2023-06-29T00:12:56.850347065Z", 20, 20]}
 9244098 |  354627 | 2023-06-29 00:12:38.926406+00 | {"IsBackFromVacation": 19}
```
In the above, it becomes player 30 (royalflush)'s turn to act at 2023-06-29 00:12:10.305587 and six and a half seconds later, she bets. That makes it player 19 (tegwin)'s turn to act and nothing happens, so twenty seconds later, player 19 is Vacationed, which causes player 19 to Fold. That makes it player 20 (jrx)'s turn to act, and two seconds later, presumably after the network catches up, player 19 is back from vacation.

Switching modems helped, but without feedback, it looks like mb2 simply has a bug where it's preemptively putting people on vacation for no reason. This could cause a player to not only abandon the platform, but to tell everyone else to stay away. As such, I'm going to re-label this issue as `high priority` even though I might not be able to work on it soon.

**Comment by ctm** [comment]
I think this was part of the culprit yesterday (#1638).  I'm marking this high priority because it really makes our software look bad. Additionally, this goes hand-in-hand with the problem that has been biting dan of late (#1627).

I probably won't allocate the time to work on this "soon", but I think the solution is to make it so check-boxes have four states, unchecked, and checked but also request-to-check-has-been-sent-but-not-acknowledged and request-to-uncheck-has-been-sent-but-not-acknowledged, perhaps 50% black for the former and 50% red for the latter.

Additionally to the two extra check-box states, true lag detection as I mention in (#1627) would help.

**Comment by ctm** [comment]
After I do the various messaging tasks, I should set Fable loose on this one. Logically, it makes sense to know which messages are responses to which messages and then decide whether we want to add a sequence number and if so, where. Since this may change the protocol, I'd rather do it sooner than later. Additionally, we can use Fable for free for around the next week or so.

**Comment by ctm** [comment]
Fable is working on this in conjunction with #1173 and #1627.

**Comment by ctm** [comment]
Deployed.
